tools/db_stress.cc

```diff
@@ -2457,28 +2465,158 @@ class StressTest {
     auto cfh = column_families_[rand_column_families[0]];
     std::unique_ptr<Iterator> iter(db_->NewIterator(readoptionscopy, cfh));
 
-    std::string key_str = Key(rand_keys[0]);
-    Slice key = key_str;
-    iter->Seek(key);
-    for (uint64_t i = 0; i < FLAGS_num_iterations && iter->Valid(); i++) {
-      if (thread->rand.OneIn(2)) {
-        iter->Next();
-      } else {
-        iter->Prev();
+    for (int64_t rkey : rand_keys) {
+      std::string key_str = Key(rkey);
+      Slice key = key_str;
+
+      if (readoptionscopy.iterate_upper_bound != nullptr &&
+          thread->rand.OneIn(2)) {
+        // 1/2 chance, change the upper bound.
+        // It is possible that it is changed without first use, but there is no
+        // problem with that.
+        int64_t rand_upper_key = GenerateOneKey(thread, FLAGS_ops_per_thread);
+        upper_bound_str = Key(rand_upper_key);
+        upper_bound = Slice(upper_bound_str);
       }
-    }
 
-    if (s.ok()) {
-      thread->stats.AddIterations(1);
-    } else {
-      thread->stats.AddErrors(1);
+      // Set up an iterator and does the same without bounds and with total
+      // order seek and compare the results. This is to identify bugs related
+      // to bounds, prefix extractor or reseeking. Sometimes we are comparing
+      // iterators with the same set-up, and it doesn't hurt to check them
+      // to be equal.
+      ReadOptions cmp_ro;
+      cmp_ro.snapshot = snapshot;
+      cmp_ro.total_order_seek = true;
+      std::unique_ptr<Iterator> cmp_iter(
+          db_->NewIterator(readoptionscopy, cfh));
+      bool diverged = false;
+
+      iter->Seek(key);
+      cmp_iter->Seek(key);
+      VerifyIterator(thread, readoptionscopy, iter.get(), cmp_iter.get(), key,
+                     &diverged);
+
+      for (uint64_t i = 0; i < FLAGS_num_iterations && iter->Valid(); i++) {
+        if (thread->rand.OneIn(2)) {
+          iter->Next();
+          if (cmp_iter->Valid()) {
```

Comment:
The if statement could be removed if add diverged to the for loop condition.

Reply:
But what's the point of complicating the code further more while the performance here is not critical.

Reply:
I found that simpler. But this is not a big deal.

Reply:
I can do that.

Reply:
I meant "diverged to the for loop condition." not to the "if condition". But it does not matter much.

Reply:
It's still valuable to do Next(), Prev() after diverged to make sure it doesn't crash, not?